src/geotop/tensor.h

```diff
@@ -0,0 +1,167 @@
+   * @param _ncl,_nch lower and upper bound for columns
+   * @param _ndl,_ndh lower and upper bound for depth
+   */
+
+    Tensor(const std::size_t _ndh,  const std::size_t _ndl,
+           const std::size_t _nrh, const std::size_t _nrl, const std::size_t _nch,  const std::size_t _ncl):
+            ndh{_ndh}, ndl{_ndl}, nrh{_nrh}, nrl{_nrl}, nch{_nch}, ncl{_ncl},
+            n_dep{ndh-ndl+1}, n_row{nrh-nrl+1}, n_col{nch-ncl+1}, co { new T[(ndh-ndl+1)*(nrh-nrl+1)*(nch-ncl+1)]{} } {}
+    // .................................................................................|-->initialize all elements to 0
+
+    Tensor(const std::size_t d, const std::size_t r, const std::size_t c): Tensor{d,1,r,1,c,1} {}
+
+    /** range-checked access operator */
+    T &at(const std::size_t k, const std::size_t i, const std::size_t j) {
+        GEO_ERROR_IN_RANGE(k, ndl, ndh);
+        GEO_ERROR_IN_RANGE(i, nrl, nrh);
+        GEO_ERROR_IN_RANGE(j, ncl, nch);
+        return (*this)[(i-nrl)*n_col + (j-ncl) + (k-ndl)*(n_row*n_col)];
+    }
+
+    /** range-checked access operator */
+    const T &at(const std::size_t k, const std::size_t i, const std::size_t j) const {
+        GEO_ERROR_IN_RANGE(k, ndl, ndh);
+        GEO_ERROR_IN_RANGE(i, nrl, nrh);
+        GEO_ERROR_IN_RANGE(j, ncl, nch);
+        return (*this)[(i-nrl)*n_col + (j-ncl) + (k-ndl)*(n_row*n_col)];
+    }
+
+    T& operator[](const std::size_t i) noexcept {
+        return co[i];
+    }
+
+    const T& operator[](const std::size_t i) const noexcept {
+        return co[i];
+    }
+
+    /**
+    * access operator. When the code is compiled in debug mode, it performes
+    * a range check. No check is done when the code is compiled in release mode.
+    */
+    T& operator()(const std::size_t k, const std::size_t i, const std::size_t j) {
+#ifndef NDEBUG
+        return at(k,i,j);
+#else
+        return (*this)[(i-nrl)*n_col + (j-ncl) + (k-ndl)*(n_row*n_col)];
+
+#endif
+    }
+
+    /**
+        * access operator. When the code is compiled in debug mode, it performes
+        * a range check. No check is done when the code is compiled in release mode.
+        */
+    const T& operator()(const std::size_t k, const std::size_t i, const std::size_t j) const
+#ifdef NDEBUG
+    noexcept
+#endif
+    {
+#ifndef NDEBUG
+        return at(k,i,j);
+#else
+        return (*this)[(i-nrl)*n_col + (j-ncl) + (k-ndl)*(n_row*n_col)];
+#endif
+    }
+
+    /** set all elements of the vector to @p v
+        * this is useful to reinizialize all the elements of the vector to zero
+        * my_Tensor = 0
+        */
+    Tensor<T> &operator=(const T v) {
+        for (auto &x : *this)
+            x = v;
+        return *this;
+    }
+
+    /**
+       * Copy constructor
+       */
+    Tensor(const Tensor<T> &t)
+            : ndl{t.ndl}, ndh{t.ndh}, nrl{t.nrl}, nrh{t.nrh}, ncl{t.ncl}, nch{t.nch},
+              n_dep{ndh-ndl+1}, n_row{nrh-nrl+1}, n_col{nch-ncl+1},
```

Comment:
why not `n_dep{t.n_dep}, n_row{t.n_row}, ...` ?

Reply:
better like you're suggesting, done!